src/core/lib/surface/init.cc

```diff
@@ -118,71 +120,85 @@ void grpc_init(void) {
-    grpc_channel_init_init();
-    grpc_core::ChannelzRegistry::Init();
-    grpc_security_pre_init();
-    grpc_core::ExecCtx::GlobalInit();
-    grpc_iomgr_init();
-    gpr_timers_global_init();
-    grpc_handshaker_factory_registry_init();
-    grpc_security_init();
-    for (i = 0; i < g_number_of_plugins; i++) {
-      if (g_all_of_the_plugins[i].init != nullptr) {
-        g_all_of_the_plugins[i].init();
+  int count;
+  do {
+    count = gpr_atm_no_barrier_load(&g_initializations);
+    if (count == 0) {
+      gpr_mu_lock(&g_init_mu);
+      if (!g_initialized) {
+        grpc_core::Fork::GlobalInit();
+        grpc_fork_handlers_auto_register();
+        gpr_time_init();
+        grpc_stats_init();
+        grpc_slice_intern_init();
+        grpc_mdctx_global_init();
+        grpc_channel_init_init();
+        grpc_core::ChannelzRegistry::Init();
+        grpc_security_pre_init();
+        grpc_core::ExecCtx::GlobalInit();
+        grpc_iomgr_init();
+        gpr_timers_global_init();
+        grpc_handshaker_factory_registry_init();
+        grpc_security_init();
+        for (i = 0; i < g_number_of_plugins; i++) {
+          if (g_all_of_the_plugins[i].init != nullptr) {
+            g_all_of_the_plugins[i].init();
+          }
+        }
+        /* register channel finalization AFTER all plugins, to ensure that it's
+         * run at the appropriate time */
+        grpc_register_security_filters();
+        register_builtin_channel_init();
+        grpc_tracer_init("GRPC_TRACE");
+        /* no more changes to channel init pipelines */
+        grpc_channel_init_finalize();
+        grpc_iomgr_start();
+        g_initialized = true;
       }
+      gpr_atm_no_barrier_fetch_add(&g_initializations, 1);
+      gpr_mu_unlock(&g_init_mu);
     }
-    /* register channel finalization AFTER all plugins, to ensure that it's run
-     * at the appropriate time */
-    grpc_register_security_filters();
-    register_builtin_channel_init();
-    grpc_tracer_init("GRPC_TRACE");
-    /* no more changes to channel init pipelines */
-    grpc_channel_init_finalize();
-    grpc_iomgr_start();
-  }
-  gpr_mu_unlock(&g_init_mu);
-
+  } while (!gpr_atm_no_barrier_cas(&g_initializations, count, count + 1));
   GRPC_API_TRACE("grpc_init(void)", 0, ());
 }
 
 void grpc_shutdown(void) {
   int i;
   GRPC_API_TRACE("grpc_shutdown(void)", 0, ());
-  gpr_mu_lock(&g_init_mu);
-  if (--g_initializations == 0) {
-    {
-      grpc_core::ExecCtx exec_ctx(0);
-      {
-        grpc_timer_manager_set_threading(
-            false);  // shutdown timer_manager thread
-        grpc_executor_shutdown();
-        for (i = g_number_of_plugins; i >= 0; i--) {
-          if (g_all_of_the_plugins[i].destroy != nullptr) {
-            g_all_of_the_plugins[i].destroy();
+  int count;
+  do {
+    count = gpr_atm_no_barrier_load(&g_initializations);
+    if (count == 1) {
+      gpr_mu_lock(&g_init_mu);
+      if (gpr_atm_no_barrier_fetch_add(&g_initializations, -1) == 1) {
```

Comment:
The decrement from 1 to 0 needs acquire semantics, because it needs to observe things other threads have done to grpc globals before calling grpc_shutdown().

Similarly, all the decrements *other* than the one from 1 to 0 need release semantics to pair up with that acquire.